auth/rbac.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright © 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/RedHatInsights/insights-results-smart-proxy/types"
+	"github.com/rs/zerolog/log"
+)
+
+type RBACClient interface {
+	IsAuthorized(token string) bool
+	IsEnforcing() bool
+}
+
+type rbacClientImpl struct {
+	uri         string
+	host        string
+	client      *http.Client
+	enforceAuth bool
+}
+
+// NewRBACClient create an RBACClient from the configuration
+func NewRBACClient(conf *RBACConfig) (RBACClient, error) {
+	url, host, err := constructRBACURL(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{}
+
+	return &rbacClientImpl{
+		url,
+		host,
+		client,
+		conf.EnforceAuth,
+	}, nil
+}
+
+func (rc *rbacClientImpl) IsEnforcing() bool {
+	return rc.enforceAuth
+}
+
+// IsAuthorized checks if an account has the correct permissions to access our resources
+func (rc *rbacClientImpl) IsAuthorized(token string) bool {
+	// baseURL := "https://console.redhat.com/api/rbac/v1" // TODO: env var, separate URL from endpoint like this: https://console.redhat.com and /api/rbac/v1/access so we can include port if needed. use helper func to create url
+	// baseURL := "https://console.stage.redhat.com/api/rbac/v1"
+
+	// Here you can add additional logic to parse the response body if needed
+	// For now, we assume that a 200 response means permissions are valid
+	permissions := rc.getPermissions(token)
+	return permissions != nil
+}
+
+func (rc *rbacClientImpl) getPermissions(identityToken string) map[string][]string {
+	acls := rc.requestAccess(rc.uri, identityToken)
+	if len(acls) > 0 {
+		permissions := aggregatePermissions(acls)
+		if len(permissions) > 0 {
+			return permissions
+		}
+		return nil
+	}
+	return nil
+}
+
+// requestAccess handles the call(s) to RBAC taking into account that the response
+// is paginated
+func (rc *rbacClientImpl) requestAccess(url, identityToken string) []types.RbacData {
+	//TODO Change return to (rbacData, err) and forward error?
```

Comment:
leftover?

Reply:
probably =). But I have two or three  TODOs in the new code that I'll handle later on